src/cpuid/src/amd/indexing.rs

```diff
@@ -21,6 +21,7 @@ impl IndexLeaf<0x8000001d> for AmdCpuid {
 
     #[inline]
     fn index_leaf<'a>(&'a self) -> Self::Output<'a> {
+        // JUSTIFICATION: There is no safe alternative.
```

Comment:
I'm not sure how this could be implemented. Could you suggest an example?

Reply:
Mh, it seems in this particular case, the `transmute_vec`  is not needed because the transmute in the `map` call already does what I proposed with my comment and transmutes to the type expected by `Leaf4` (I removed it and everything still compiled). Could you double check?

Reply:
Although, wrt `.map(|(_, v)| transmute::<_, &'a Leaf8000001dSubleaf>(&v.result))`, why is this safe? Doesn't it "upcast" the lifetime (e.g. `v.result` is only borrowed for the duration of the closure, and transmute upcasts it to `'a`)? Because in that case, this code (or well, the impl for `&mut` at least) would violate Rust's aliasing rules, as later on `v.result` could get borrowed mutably a second time (because the borrow checker doesn't understand `transmute`, so for all it knows, `v.result` does not have any live references to it after the closure exits)

Reply:
The lifetime `'a` as defined on `index_leaf<'a>(&'a self) -> Self::Output<'a>` informs the borrow checker that `Self::Output` immutably borrows `self`. This prevents dropping or mutating `self` before dropping `Self::Output`.
E.g.
```rust

fn main() {
    let mut a: Vec<i32> = vec![1,2,3,4];
    let b = tester(&a);

    // drop(a);
    // drop(b); // Attempting to drop b after a errors.

    // a[0] += 2;
    // drop(b); // Attempting to mutably borrow a before dropping b errors.
}

#[derive(Debug)]
struct Wrapper<'a>(Vec<&'a u32>);

fn tester<'a>(vec: &'a Vec<i32>) -> Wrapper<'a> {
    Wrapper(vec.iter().map(|x: &'a i32|unsafe { std::mem::transmute::<&'a i32,&'a u32>(x) }).collect::<Vec<_>>())
}
```

Reply:
Ah, I see, yes you are right, nothing bad can happen outside of `index_leaf`. However, for inside the method, the lifetime annotation on the closure argument that you included in your example is important, to make sure that the iterator does not borrow for a lifetime narrower than `'a`. For example, the following compiles and is UB because `push` might reallocate:

```rs
struct Wrapper<'a>(Vec<&'a mut u32>);

    fn tester<'a>(vec: &'a mut Vec<i32>) -> Wrapper<'a> {
      let wrapper = Wrapper(
          vec.iter_mut()
              .map(|x| unsafe { std::mem::transmute::<_, &'a mut u32>(x) })
              .collect::<Vec<_>>(),
      );

      vec.push(3);

      wrapper
    }
```

Could you please add that, and also expand the safety comment with what you explained in this thread?

Reply:
I would want to avoid an lengthy comment duplicated in many places where `Index` and `IndexMut` are implemented. What comment would you suggest here?

Reply:
I think having the lengthy comment once in a prominent place (first instance of transmute in indexing.rs maybe?) and then having each other instance explicitly refer to that comment would be fine by me.

Reply:
I've added a comment on `index_leaf!`:

```text
/// Convenience macro for indexing leaves.
/// 
/// # Safety
///
/// The lifetime 'a as defined on `index_leaf<'a>(&'a self) -> Self::Output<'a>` informs the borrow
/// checker that `Self::Output` immutably borrows `self`. This prevents dropping or mutating `self`
/// before dropping `Self::Output`.
///
/// E.g.
/// ```ignore
/// fn main() {
///     let mut a: Vec<i32> = vec![1, 2, 3, 4];
///     let b = tester(&a);
/// 
///     // drop(a);
///     // drop(b); // Attempting to drop b after a errors.
/// 
///     // a[0] += 2;
///     // drop(b); // Attempting to mutably borrow a before dropping b errors.
/// }
/// 
/// #[derive(Debug)]
/// struct Wrapper<'a>(Vec<&'a u32>);
/// 
/// fn tester<'a>(vec: &'a Vec<i32>) -> Wrapper<'a> {
///     Wrapper(
///         vec.iter()
///             .map(|x: &'a i32| unsafe { std::mem::transmute::<&'a i32, &'a u32>(x) })
///             .collect::<Vec<_>>(),
///     )
/// }
/// ```
```

and comments on the transmutes like:
```
// SAFETY: Transmuting references to same size and alignment types is safe. For
// further information See `index_leaf!`.
```

Reply:
Sounds good! Please also make sure that the iterator borrows for `self` for `'a` instead of a local lifetime, just to be extra safe, and extend the justification comment with why we use `transmute` over returning a copy.

Reply:
> and extend the justification comment with why we use transmute over returning a copy.

This comment would be on returning a reference is preferable to returning a copy. I'm not sure this is specific to this implementation.

Reply:
> > Please also make sure that the iterator borrows for self for 'a instead of a local lifetime, just to be extra safe
> 
> Are you suggesting adding `: &'a` and `: &'a mut` respectively to the map in all the implementations?

Yep

> > and extend the justification comment with why we use transmute over returning a copy.
> 
> This comment would be on why returning a reference is preferable to returning a copy. I'm not sure this is specific to this implementation. Similar purpose to a generic getter.

Since here the fact that we return a reference is what makes us required `unsafe`, we should properly motivate it, imo.

Reply:
> Yep

I would consider this a nit. I think the level function lifetime is clear.

> Since here the fact that we return a reference is what makes us required unsafe, we should properly motivate it, imo.

It would require unsafe to return a copy, as we would need to transmute into the leaf then copy it.

Reply:
> > Yep
> 
> I would consider this a nit. I think the level function lifetime is clear.

Without the annotation, the compiler infers the lifetime for which the iterator borrows `self.0`, and it might infer it to be narrower than `'a`. In this case, the transmute upcasts, and for the duration of the method safe code _could_ result in UB, see my example a few comments above. 

> > Since here the fact that we return a reference is what makes us required unsafe, we should properly motivate it, imo.
> 
> It would require unsafe to return a copy, as we would need to transmute into the leaf then copy it.

Wouldn't it be just `.map(|&entry| Leaf8000001dSubleaf {eax: entry.result.eax, /* and so on* /})`? Where would we still need a transmute? (however, if I'm overlooking something and we do need the transmute anyway, then disregard what I said about adding a comment)

Reply:
> Wouldn't it be just .map(|&entry| Leaf8000001dSubleaf {eax: Leaf8000001dEax(entry.result.eax), /* and so on* /})? Where would we still need a transmute? (however, if I'm overlooking something and we do need the transmute anyway, then disregard what I said about adding a comment)

You are correct it can be done without unsafe.

For

> the justification comment with why we use transmute over returning a copy.

What comment would you suggest?

Reply:
I've added the comment:

> get  and get_mut  is a common Rust pattern, these functions typically return &T  and &mut T  respectively. In the case of get it is preferable to return a reference over a copy, a reference can be optimized to a copy (while a copy cannot be optimized to a reference). To return a reference in either case requires std::mem::transmute (or casting the reference as a pointer, to the same affect). This is why std::mem::transmute  is used here.

To the `index_leaf!` macro.